presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java

```diff
@@ -0,0 +1,332 @@
+        this.filter = requireNonNull(filter, "filter is null").orElse(null);
+        this.outputRequired = outputRequired;
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+    }
+
+    @Override
+    public void startStripe(InputStreamSources dictionaryStreamSources, List<ColumnEncoding> encoding)
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+        dataStreamSource = missingStreamSource(BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+        dataStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, DATA, BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .addValue(streamDescriptor)
+                .toString();
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(values) + sizeOf(nulls) + sizeOf(outputPositions);
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        boolean nullsAllowed = filter == null || filter.testNull();
+        boolean recordNulls = nullsAllowed && presentStream != null;
+        if (outputRequired) {
+            ensureValuesCapacity(positionCount, recordNulls);
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
+        }
+        else {
+            outputPositions = positions;
+        }
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
+        allNulls = false;
+        if (dataStream == null && presentStream != null) {
+            // all nulls
+            presentStream.skip(positions[positionCount - 1]);
+            streamPosition = positions[positionCount - 1] + 1;
+
+            if (nullsAllowed) {
+                outputPositionCount = positionCount;
+                allNulls = true;
+            }
+            else {
+                outputPositionCount = 0;
+            }
+        }
```

Comment:
We might need to check
```java
else if (dataStream == null) {
    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
}

Reply:
dataStream is null in this case... did you think of some other branch?